### ISSUE TITLE
Failed to build v2021.3.0 with GCC 11 and glibc 2.34 on ppc64le

### ISSUE THREAD

**Issue by mscastanho** [oneapi-src/oneTBB#530]
Hi, I've been facing several issues when trying to build the latest oneTBB release on ppc64le with GCC 11 + glibc 2.34. I have workarounds for some, and I'm ignoring others for now with the appropriate `-Wno-error=*` flags. 

I'm building with:
```
~/build/tbb$ LDFLAGS=-m64 CXXFLAGS='-m64' CC=gcc-11 CXX=g++-11 cmake ~/src/tbb/ && VERBOSE=1 cmake --build . -j10 -- arch=ppc64le
```

Errors when building with GCC 11.2 and glibc 2.34:

1. Build errors related to `getSmallObjectIndex`
    Fixed by cherry-picking:
   fa944e19 Remove arch macros checks for getSmallObjectIndex (#461)

2. Use of deprecated mallinfo
   Can be fixed by the following diff, which is somewhat similar to PR #438 (but the PR is more complete).
```diff
diff --git a/test/tbbmalloc/test_malloc_overload.cpp b/test/tbbmalloc/test_malloc_overload.cpp
index 35dd7b26..29fab0c0 100644
--- a/test/tbbmalloc/test_malloc_overload.cpp
+++ b/test/tbbmalloc/test_malloc_overload.cpp
@@ -55,6 +55,7 @@
 #if __GLIBC__
   // aligned_alloc available since GLIBC 2.16
   #define __TBB_ALIGNED_ALLOC_PRESENT __GLIBC_PREREQ(2, 16)
+  #define __TBB_MALLINFO2_PRESENT __GLIBC_PREREQ(2, 33)
 #endif // __GLIBC__
  // later Android doesn't have valloc or dlmalloc_usable_size
 #define __TBB_VALLOC_PRESENT (__linux__ && __ANDROID_API__<21) || __APPLE__
@@ -448,7 +449,11 @@ TEST_CASE("Main set of tests") {
     CheckMemalignFuncOverload(aligned_alloc, free);
 #endif

+#ifdef __TBB_MALLINFO2_PRESENT
+    struct mallinfo2 info = mallinfo2();
+#else
     struct mallinfo info = mallinfo();
+#endif
     // right now mallinfo initialized by zero
     REQUIRE((!info.arena && !info.ordblks && !info.smblks && !info.hblks
            && !info.hblkhd && !info.usmblks && !info.fsmblks
```

3. `array-bounds` warnings

   I found these a bit harder to debug, do you know if these are actually  false-positives or not? If they turn out to be false-positives, I can report those to GCC instead.

```
[ 45%] Built target test_join_node_msg_key_matching
In file included from /home/mscastanho/src/tbb/src/tbb/../../include/tbb/../oneapi/tbb/concurrent_unordered_map.h:21,
                 from /home/mscastanho/src/tbb/src/tbb/../../include/tbb/concurrent_unordered_map.h:17,
                 from /home/mscastanho/src/tbb/test/tbb/test_concurrent_unordered_map.cpp:22:
/home/mscastanho/src/tbb/src/tbb/../../include/tbb/../oneapi/tbb/detail/_concurrent_unordered_base.h: In member function ‘void tbb::detail::d1::concurrent_unordered_base<Traits>::init_bucket(tbb::detail::d1::concurrent_unordered_base<Traits>::size_type) [with Traits = tbb::detail::d1::concurrent_unordered_map_traits<int, int, std::hash<int>, std::equal_to<int>, NotAlwaysEqualAllocator<std::pair<const int, int> >, false>]’:
/home/mscastanho/src/tbb/src/tbb/../../include/tbb/../oneapi/tbb/detail/_concurrent_unordered_base.h:172:20: warning: array subscript ‘tbb::detail::d1::value_node<std::pair<const int, int>, long unsigned int>[0]’ is partly outside array bounds of ‘unsigned char [16]’ [-Warray-bounds]
  172 |     ~value_node() {}
      |                    ^
In file included from /usr/include/c++/11/powerpc64le-suse-linux/bits/c++allocator.h:33,
                 from /usr/include/c++/11/bits/allocator.h:46,
                 from /usr/include/c++/11/memory:64,
                 from /home/mscastanho/src/tbb/src/tbb/../../include/tbb/../oneapi/tbb/detail/_template_helpers.h:27,
                 from /home/mscastanho/src/tbb/src/tbb/../../include/tbb/../oneapi/tbb/detail/_containers_helpers.h:20,
                 from /home/mscastanho/src/tbb/src/tbb/../../include/tbb/../oneapi/tbb/detail/_concurrent_unordered_base.h:25,
                 from /home/mscastanho/src/tbb/src/tbb/../../include/tbb/../oneapi/tbb/concurrent_unordered_map.h:21,
                 from /home/mscastanho/src/tbb/src/tbb/../../include/tbb/concurrent_unordered_map.h:17,
                 from /home/mscastanho/src/tbb/test/tbb/test_concurrent_unordered_map.cpp:22:
/usr/include/c++/11/ext/new_allocator.h:127:48: note: referencing an object of size 16 allocated by ‘void* operator new(std::size_t)’
  127 |         return static_cast<_Tp*>(::operator new(__n * sizeof(_Tp)));
      |                                  ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
In file included from /home/mscastanho/src/tbb/src/tbb/../../include/tbb/../oneapi/tbb/concurrent_unordered_map.h:21,
                 from /home/mscastanho/src/tbb/src/tbb/../../include/tbb/concurrent_unordered_map.h:17,
                 from /home/mscastanho/src/tbb/test/tbb/test_concurrent_unordered_map.cpp:22:
/home/mscastanho/src/tbb/src/tbb/../../include/tbb/../oneapi/tbb/detail/_concurrent_unordered_base.h: In member function ‘void tbb::detail::d1::concurrent_unordered_base<Traits>::init_bucket(tbb::detail::d1::concurrent_unordered_base<Traits>::size_type) [with Traits = tbb::detail::d1::concurrent_unordered_map_traits<int, int, std::hash<int>, std::equal_to<int>, NotAlwaysEqualAllocator<std::pair<const int, int> >, true>]’:
/home/mscastanho/src/tbb/src/tbb/../../include/tbb/../oneapi/tbb/detail/_concurrent_unordered_base.h:172:20: warning: array subscript ‘tbb::detail::d1::value_node<std::pair<const int, int>, long unsigned int>[0]’ is partly outside array bounds of ‘unsigned char [16]’ [-Warray-bounds]
  172 |     ~value_node() {}
      |                    ^
```

4. `sizeof-array-div` warning

This is a new warning type on GCC 11.

```
[ 45%] Building CXX object test/CMakeFiles/conformance_split_node.dir/conformance/conformance_split_node.cpp.o
cd /home/mscastanho/build/tbb/test && /usr/bin/g++-11  -I/home/mscastanho/src/tbb/test/.. -I/home/mscastanho/src/tbb/test -I/home/mscastanho/src/tbb/src/tbb/../../include -m64 -O2 -g -DNDEBUG -Wall -Wextra -Werror -Wfatal-errors -Wno-error=array-bounds -Wshadow -Wcast-qual -Woverloaded-virtual -Wnon-virtual-dtor -pthread -std=c++11 -MD -MT test/CMakeFiles/conformance_split_node.dir/conformance/conformance_split_node.cpp.o -MF CMakeFiles/conformance_split_node.dir/conformance/conformance_split_node.cpp.o.d -o CMakeFiles/conformance_split_node.dir/conformance/conformance_split_node.cpp.o -c /home/mscastanho/src/tbb/test/conformance/conformance_split_node.cpp
In file included from /home/mscastanho/src/tbb/src/tbb/../../include/tbb/../oneapi/tbb/flow_graph.h:405,
                 from /home/mscastanho/src/tbb/src/tbb/../../include/tbb/flow_graph.h:17,
                 from /home/mscastanho/src/tbb/test/tbb/test_tagged_msg.cpp:21:
/home/mscastanho/src/tbb/src/tbb/../../include/tbb/../oneapi/tbb/detail/_flow_graph_types_impl.h: In instantiation of ‘const size_t tbb::detail::d1::Wrapper<std::vector<int> [2]>::space_size’:
/home/mscastanho/src/tbb/src/tbb/../../include/tbb/../oneapi/tbb/detail/_flow_graph_types_impl.h:173:22:   required from ‘struct tbb::detail::d1::Wrapper<std::vector<int> [2]>’
/home/mscastanho/src/tbb/test/tbb/test_tagged_msg.cpp:90:35:   required from here
/home/mscastanho/src/tbb/src/tbb/../../include/tbb/../oneapi/tbb/detail/_flow_graph_types_impl.h:172:56: error: expression does not compute the number of elements in this array; element type is ‘std::vector<int>’, not ‘char’ [-Werror=sizeof-array-div]
  172 |     static const size_t space_size = sizeof(ArrayType) / sizeof(char);
      |                                      ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
compilation terminated due to -Wfatal-errors.
```

**Comment by alexey-katranov** [comment]
2 & 4 is fixed in #433 
3 is fixed in #447

**Comment by mscastanho** [comment]
Awesome! With those fixes I'm able to fix the issues. Thanks!
